File: backend/prompts.py
from __future__ import annotations

import json
from pathlib import Path

from .paths import DEFAULT_PROMPTS_PATH

PROMPT_KEYS: tuple[str, ...] = (
    "brief_prompt",
    "dossier_prompt",
    "post_draft_prompt",
    "mod_review_prompt",
    "revise_prompt",
    "native_polish_prompt",
    "engagement_prompt",
)

REQUIRED_PLACEHOLDERS: dict[str, tuple[str, ...]] = {
    "brief_prompt": ("{{pre_materials}}",),
    "dossier_prompt": ("{{subreddit_name}}", "{{subreddit_meta}}", "{{subreddit_rules}}", "{{corpus_excerpt}}"),
    # post_draft_prompt placeholders are OPTIONAL to support "client-provided draft" workflows
    # where we don't want to spend tokens injecting large context blocks.
    "post_draft_prompt": (),
    "mod_review_prompt": (
        "{{subreddit_name}}",
        "{{current_date}}",
        "{{subreddit_rules}}",
        "{{subreddit_dossier}}",
        "{{corpus_excerpt}}",
        "{{post_draft}}",
    ),
    "revise_prompt": ("{{subreddit_name}}", "{{current_date}}", "{{mod_review}}", "{{post_draft}}"),
    "native_polish_prompt": ("{{subreddit_name}}", "{{current_date}}", "{{subreddit_dossier}}", "{{post_revision}}"),
    "engagement_prompt": (
        "{{subreddit_name}}",
        "{{current_date}}",
        "{{subreddit_dossier}}",
        "{{corpus_excerpt}}",
        "{{post_final}}",
    ),
}


def load_prompts_file(path: Path) -> dict[str, str]:
    raw = json.loads(path.read_text(encoding="utf-8"))
    if not isinstance(raw, dict):
        raise ValueError("Prompts file must be a JSON object.")

    prompts: dict[str, str] = {}
    for key, value in raw.items():
        if not isinstance(key, str) or not isinstance(value, str):
            continue
        prompts[key] = value
    return prompts


def load_default_prompts(path: Path = DEFAULT_PROMPTS_PATH) -> dict[str, str]:
    prompts = load_prompts_file(path)

    missing = [k for k in PROMPT_KEYS if k not in prompts]
    if missing:
        raise ValueError(f"Default prompts file missing keys: {missing}")

    return {k: prompts[k] for k in PROMPT_KEYS}


def validate_prompts(prompts: dict[str, str], *, skip_keys: set[str] | None = None) -> None:
    skip = skip_keys or set()

    for key in PROMPT_KEYS:
        if key in skip:
            continue
        value = prompts.get(key)
        if not isinstance(value, str) or not value.strip():
            raise ValueError(f"Prompt '{key}' must be a non-empty string.")

    for key, required_tokens in REQUIRED_PLACEHOLDERS.items():
        if key in skip:
            continue
        value = prompts.get(key) or ""
        for token in required_tokens:
            if token not in value:
                raise ValueError(f"Prompt '{key}' must contain placeholder {token}.")


def merge_prompts(
    default_prompts: dict[str, str],
    overrides: dict[str, str] | None,
    *,
    skip_keys: set[str] | None = None,
) -> dict[str, str]:
    merged = dict(default_prompts)
    if overrides:
        for key, value in overrides.items():
            if key not in PROMPT_KEYS:
                raise ValueError(f"Unknown prompt key: {key}")
            if not isinstance(value, str):
                raise ValueError(f"Prompt override '{key}' must be a string.")
            merged[key] = value
    validate_prompts(merged, skip_keys=skip_keys)
    return merged


def write_prompts_file(path: Path, prompts: dict[str, str]) -> None:
    path.write_text(json.dumps(prompts, ensure_ascii=False, indent=2), encoding="utf-8")

File: frontend/src/prompts.ts
export type PromptKey =
  | 'brief_prompt'
  | 'dossier_prompt'
  | 'post_draft_prompt'
  | 'mod_review_prompt'
  | 'revise_prompt'
  | 'native_polish_prompt'
  | 'engagement_prompt'

export const PROMPT_KEYS: PromptKey[] = [
  'brief_prompt',
  'dossier_prompt',
  'post_draft_prompt',
  'mod_review_prompt',
  'revise_prompt',
  'native_polish_prompt',
  'engagement_prompt',
]

export type PromptMeta = {
  key: PromptKey
  title: string
  description: string
  requiredPlaceholders: string[]
}

export const PROMPT_METAS: PromptMeta[] = [
  {
    key: 'brief_prompt',
    title: '阶段 0：产品 Brief 抽取',
    description:
      '从前置资料中抽取“可执行的产品 brief”（不应引用原文）。必须包含占位符。',
    requiredPlaceholders: ['{{pre_materials}}'],
  },
  {
    key: 'dossier_prompt',
    title: '阶段 1：Subreddit Dossier（风格/规则/禁忌）',
    description:
      '基于抓取到的 meta + rules + 语料片段，生成该 sub 的写作/互动指南。',
    requiredPlaceholders: [
      '{{subreddit_name}}',
      '{{subreddit_meta}}',
      '{{subreddit_rules}}',
      '{{corpus_excerpt}}',
    ],
  },
  {
    key: 'post_draft_prompt',
    title: '阶段 2：Post 初稿（v1）',
    description: '基于产品 brief + dossier，生成单一最佳版本的 post 初稿。',
    requiredPlaceholders: [
      '{{subreddit_name}}',
      '{{product_brief}}',
      '{{subreddit_dossier}}',
    ],
  },
  {
    key: 'mod_review_prompt',
    title: '阶段 3：Mod 审核（Pass/Fail）',
    description: '站在版主视角逐条对照规则，给出是否会删帖与改进建议。',
    requiredPlaceholders: [
      '{{subreddit_name}}',
      '{{subreddit_rules}}',
      '{{subreddit_dossier}}',
      '{{post_draft}}',
    ],
  },
  {
    key: 'revise_prompt',
    title: '阶段 4：合规修订（v2）',
    description: '根据 mod 审核结果，输出更安全的 v2 版本。',
    requiredPlaceholders: ['{{subreddit_name}}', '{{mod_review}}', '{{post_draft}}'],
  },
  {
    key: 'native_polish_prompt',
    title: '阶段 5：Reddit-native 打磨（Final）',
    description: '去营销味、增强真实感与互动钩子（不刻意制造错字）。',
    requiredPlaceholders: [
      '{{subreddit_name}}',
      '{{subreddit_dossier}}',
      '{{post_revision}}',
    ],
  },
  {
    key: 'engagement_prompt',
    title: '阶段 6：互动文案包（OP-only）',
    description:
      '生成 OP 可用的首评/补充评论 + 回复模板（不伪装其他用户）。',
    requiredPlaceholders: [
      '{{subreddit_name}}',
      '{{subreddit_dossier}}',
      '{{post_final}}',
    ],
  },
]

export function diffPrompts(
  defaults: Record<string, string>,
  current: Record<string, string>,
): Record<string, string> {
  const overrides: Record<string, string> = {}
  for (const key of PROMPT_KEYS) {
    if ((current[key] ?? '').trim() !== (defaults[key] ?? '').trim()) {
      overrides[key] = current[key] ?? ''
    }
  }
  return overrides
}

export function validatePrompts(
  prompts: Record<string, string>,
): Record<PromptKey, string | null> {
  const errors = {} as Record<PromptKey, string | null>
  for (const meta of PROMPT_METAS) {
    const value = (prompts[meta.key] ?? '').trim()
    if (!value) {
      errors[meta.key] = '不能为空'
      continue
    }
    const missing = meta.requiredPlaceholders.filter((p) => !value.includes(p))
    if (missing.length > 0) {
      errors[meta.key] = `缺少占位符：${missing.join(', ')}`
      continue
    }
    errors[meta.key] = null
  }
  return errors
}

